src/main/java/seedu/address/model/module/UniqueModuleList.java

```diff
@@ -0,0 +1,137 @@
+
+import java.util.Iterator;
+import java.util.List;
+
+import javafx.collections.FXCollections;
+import javafx.collections.ObservableList;
+import seedu.address.model.module.exceptions.DuplicateModuleException;
+import seedu.address.model.module.exceptions.ModuleNotFoundException;
+
+/**
+ * A list of modules that enforces uniqueness between its elements and does not allow nulls.
+ * A module is considered unique by comparing using {@code Module#isSameModule(Module)}. As such, adding and updating of
+ * modules uses Module#isSameModule(Module) for equality so as to ensure that the module being added or updated is
+ * unique in terms of identity in the UniqueModuleList. However, the removal of a module uses Module#equals(Object) so
+ * as to ensure that the module with exactly the same fields will be removed.
+ *
+ * Supports a minimal set of list operations.
+ *
+ * @see Module#isSameModule(Module)
+ */
+public class UniqueModuleList implements Iterable<Module> {
+
+    private final ObservableList<Module> internalList = FXCollections.observableArrayList();
+    private final ObservableList<Module> internalUnmodifiableList =
+            FXCollections.unmodifiableObservableList(internalList);
+
+    /**
+     * Returns true if the list contains an equivalent module as the given argument.
+     */
+    public boolean contains(Module toCheck) {
+        requireNonNull(toCheck);
+        return internalList.stream().anyMatch(toCheck::isSameModule);
+    }
+
+    /**
+     * Adds a module to the list.
+     * The module must not already exist in the list.
+     */
+    public void add(Module toAdd) {
+        requireNonNull(toAdd);
+        if (contains(toAdd)) {
+            throw new DuplicateModuleException();
+        }
+        internalList.add(toAdd);
+    }
+
+    /**
+     * Replaces the module {@code target} in the list with {@code editedModule}.
+     * {@code target} must exist in the list.
+     * The module identity of {@code editedModule} must not be the same as another existing module in the list.
+     */
+    public void setModule(Module target, Module editedModule) {
+        requireAllNonNull(target, editedModule);
+
+        int index = internalList.indexOf(target);
+        if (index == -1) {
+            throw new ModuleNotFoundException();
+        }
+
+        if (!target.isSameModule(editedModule) && contains(editedModule)) {
+            throw new DuplicateModuleException();
+        }
+
+        internalList.set(index, editedModule);
+    }
+
+    /**
+     * Removes the equivalent module from the list.
+     * The module must exist in the list.
+     */
+    public void remove(Module toRemove) {
+        requireNonNull(toRemove);
+        if (!internalList.remove(toRemove)) {
+            throw new ModuleNotFoundException();
+        }
+    }
+
+    public void setModules(UniqueModuleList replacement) {
+        requireNonNull(replacement);
+        internalList.setAll(replacement.internalList);
```

Comment:
I am not sure here. For internalList, it is a final attribute of UniqueModuleList class. Maybe adding a public method, e.g. getInternalList(), will be better? 
(e.g. 
```replacement.internalList```
changes to 
```replacement.getInternalList()```
)